runtime/commonMain/src/kotlinx/serialization/json/JsonExceptions.kt

```diff
@@ -28,22 +28,30 @@ internal class JsonEncodingException(message: String) : JsonException(message)
 internal fun JsonDecodingException(offset: Int, message: String, input: String) =
     JsonDecodingException(offset, "$message.\n JSON input: ${input.minify(offset)}")
 
-internal fun InvalidFloatingPoint(value: Number, type: String, output: String) = JsonEncodingException(
-    "'$value' is not a valid '$type' as per JSON specification. " +
-            "You can enable 'serializeSpecialFloatingPointValues' property to serialize such values\n" +
+internal fun InvalidFloatingPoint(value: Number, output: String) = JsonEncodingException(
+    "Unexpected special floating-point value $value. By default, " +
+            "non-finite floating point values are prohibited because they do not conform JSON specification. " +
+            "It is possible to serialize them using 'JsonBuilder.allowSpecialFloatingPointValues = true'\n" +
             "Current output: ${output.minify()}"
 )
 
-internal fun InvalidFloatingPoint(value: Number, key: String, type: String, output: String) = JsonEncodingException(
-    "'$value' with key '$key' is not a valid $type as per JSON specification. " +
-            "You can enable 'serializeSpecialFloatingPointValues' property to serialize such values.\n" +
+internal fun InvalidFloatingPointDecoded(value: Number, key: String, output: String) =
+    JsonDecodingException(-1, invalidFp(value, key, output))
+
+internal fun InvalidFloatingPoint(value: Number, key: String, output: String) =
+    JsonEncodingException(invalidFp(value, key, output))
+
+private fun invalidFp(value: Number, key: String, output: String): String {
```

Comment:
This confuses me a lot. I've counted three different ways to throw an exception with 3 different messages: `InvalidFloatingPoint`, `InvalidFloatingPointDecoded`, and `reader.fail` (and strange `invalidFp`). I don't get the logic why you choose one or another. Maybe it's possible to collect them all in one place and name `InvalidFloatingPointEncoded` and `InvalidFloatingPointDecoded`? (with all necessary overloads)

Reply:
Sure, done